### PR TITLE
chore: disallow insert values for ROWPARTITION and ROWOFFSET

### PR DESCRIPTION
**Pull request by Sullivan-Patrick**
### Description 
This change is part of [KLIP-50](https://github.com/confluentinc/ksql/blob/master/design-proposals/klip-50-partition-and-offset-in-ksqldb.md). It disallows the use of `INSERT VALUES` into `ROWPARTITION` and `ROWOFFSET` pseudocolumns. The implementation allows for user columns with these names to continue to allow for inserts (this situation can occur iff a user created a stream or table which contained a user column named `ROWPARTITION`/`ROWOFFSET` before upgrading to a server version with the pseudocolumns enabled),

To make this easier to write and test, it's based on top of the still-unmerged branch associated with https://github.com/confluentinc/ksql/pull/8120. This change only concerns commits after 2ae058a, inclusive.

### Testing done 
New QTTs to verify behavior with and without feature enabled.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/ExecutionStepFactory.java [hunk 1]
```diff
@@ -149,7 +150,7 @@ public static TableSource tableSource(
         formats,
         timestampColumn,
         sourceSchema,
-        SystemColumns.CURRENT_PSEUDOCOLUMN_VERSION_NUMBER,
+        pseudoColumnVersion,
```
Comment (Sullivan-Patrick) [inline ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/ExecutionStepFactory.java:153]: Ignore this change, included in this commit due to my lack of proficiency in Git.

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/row-offset-and-partition.json [hunk 2]
```diff
@@ -134,12 +134,142 @@
...
+        "CREATE STREAM input (id INT KEY, val INTEGER) WITH (kafka_topic='test_topic', value_format='JSON');",
+        "CREATE STREAM output AS SELECT ROWOFFSET FROM input EMIT CHANGES;"
+      ],
+      "expectedException": {
+        "type": "io.confluent.ksql.util.KsqlStatementException",
+        "message": "SELECT column 'ROWOFFSET' cannot be resolved."
+      }
+    },
+    {
+      "name": "verify INSERT VALUES works for ROWPARTITION and ROWOFFSET with feature disabled (matching types)",
```
Comment (vcrfxia) [inline ksqldb-functional-tests/src/test/resources/query-validation-tests/row-offset-and-partition.json:173]: This is not the right place to be testing INSERT VALUES: QTT does not support INSERT VALUES. Rather, INSERT VALUES is tested in RQTT instead (see the file named `insert-values.json`).

Also, I don't think we need tests for verifying that inserts into ROWPARTITION AND ROWOFFSET are not allowed if the feature flag is not enabled. If there were a bug here, it'd be because the feature flag is not set up properly. This is a good thing to test while introducing the feature flag, but it's not necessary to check in since we'd just remove it once we ungate the feature flag anyway.

File: ksqldb-engine/src/main/java/io/confluent/ksql/engine/generic/GenericRecordFactory.java [hunk 3]
```diff
@@ -78,7 +78,13 @@ public KsqlGenericRecord build(
...
     for (ColumnName col : columns) {
       if (!schemaWithRowTime.findColumn(col).isPresent()) {
-        throw new KsqlException("Column name " + col + " does not exist.");
+        final String message;
+        if (SystemColumns.disallowedForInsertValues(col)) {
+          message = "Inserting into column " + col + " is not allowed.";
+        } else {
+          message = "Column name " + col + " does not exist.";
+        }
+        throw new KsqlException(message);
```
Comment (vcrfxia) [inline ksqldb-engine/src/main/java/io/confluent/ksql/engine/generic/GenericRecordFactory.java:87]: The changes in this file should be accompanied by unit test changes. See GenericRecordFactoryTest.java and InsertValuesExecutorTest.java for inspiration.

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/row-offset-and-partition.json (hunk 2 above)
Comment (vcrfxia) [reply ksqldb-functional-tests/src/test/resources/query-validation-tests/row-offset-and-partition.json:173]: Incidentally, QTT _is_ the correct place to check that inserting into ROWPARTITION and ROWOFFSET are disabled for INSERT INTO statements (in contrast to INSERT VALUES statements). It doesn't have to be in this PR, but we should make sure that ROWPARTITION and ROWOFFSET are properly disabled for INSERT INTO, and add the relevant test coverage. (Can you check that we have a ticket capturing this, and add one if not?)

**Review by vcrfxia — COMMENTED** [review]
Thanks @Sullivan-Patrick -- high-level approach looks good. Comments inline mostly around test coverage.

File: ksqldb-common/src/main/java/io/confluent/ksql/schema/ksql/SystemColumns.java [hunk 4]
```diff
@@ -165,6 +168,22 @@ public static boolean mustBeMaterializedForTableJoins(final ColumnName columnNam
         .mustBeMaterializedForTableJoins;
   }
 
+  /**
+   * Checks if a given ColumnName is associated with a pseudo column that is disallowed
```
Comment (vcrfxia) [inline ksqldb-common/src/main/java/io/confluent/ksql/schema/ksql/SystemColumns.java:172]:
```suggestion
   * Checks if a given pseudo column name is associated with a pseudo column that is disallowed
```

File: ksqldb-engine/src/main/java/io/confluent/ksql/engine/generic/GenericRecordFactory.java [hunk 5]
```diff
@@ -75,23 +75,29 @@ public KsqlGenericRecord build(
...
-    final LogicalSchema schemaWithRowTime = withRowTime(schema);
+    final LogicalSchema schemaWithPseudoColumns = withPseudoColumns(schema, config);
     for (ColumnName col : columns) {
-      if (!schemaWithRowTime.findColumn(col).isPresent()) {
+
+      if (!schemaWithPseudoColumns.findColumn(col).isPresent()) {
         throw new KsqlException("Column name " + col + " does not exist.");
       }
+
+      if (SystemColumns.isSystemColumn(col, config)
```
Comment (vcrfxia) [inline ksqldb-engine/src/main/java/io/confluent/ksql/engine/generic/GenericRecordFactory.java:85]: Don't we need to check that the column is a pseudo column, not just a system column? I think confusion like this is why we should simply have `isDisallowedForInsertValues()` accept any column name rather than throw on non-pseudo-columns.

File: ksqldb-engine/src/test/java/io/confluent/ksql/engine/generic/GenericRecordFactoryTest.java [hunk 6]
```diff
@@ -43,21 +43,35 @@
...
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoJUnitRunner;
+import org.mockito.junit.MockitoRule;
 
+@RunWith(MockitoJUnitRunner.class)
 public class GenericRecordFactoryTest {
 
+  @Rule public MockitoRule rule = MockitoJUnit.rule();
```
Comment (vcrfxia) [inline ksqldb-engine/src/test/java/io/confluent/ksql/engine/generic/GenericRecordFactoryTest.java:57]: This looks redundant?

File: ksqldb-engine/src/test/java/io/confluent/ksql/engine/generic/GenericRecordFactoryTest.java [hunk 7]
```diff
@@ -43,21 +43,35 @@
...
 
   private final AtomicLong clock = new AtomicLong();
-  private final KsqlConfig config = new KsqlConfig(ImmutableMap.of());
   private final FunctionRegistry functions = TestFunctionRegistry.INSTANCE.get();
-  private final GenericRecordFactory recordFactory = new GenericRecordFactory(config, functions, clock::get);
+
+  @Mock
+  private KsqlConfig ksqlConfig;
+
+  //fixme: why is config not mocked yet here?
```
Comment (vcrfxia) [inline ksqldb-engine/src/test/java/io/confluent/ksql/engine/generic/GenericRecordFactoryTest.java:69]: Remove?

File: ksqldb-engine/src/test/java/io/confluent/ksql/engine/generic/GenericRecordFactoryTest.java [hunk 8]
```diff
@@ -43,21 +43,35 @@
...
+  private GenericRecordFactory recordFactory;
 
   @Before
   public void setUp() {
+    recordFactory = new GenericRecordFactory(ksqlConfig, functions, clock::get);
```
Comment (vcrfxia) [inline ksqldb-engine/src/test/java/io/confluent/ksql/engine/generic/GenericRecordFactoryTest.java:74]: It's customary to have this be the last line in `setUp()`, since this is initializing the variable being tested (so typically all the other setup should happen first).

File: ksqldb-engine/src/test/java/io/confluent/ksql/engine/generic/GenericRecordFactoryTest.java [hunk 9]
```diff
@@ -350,4 +364,44 @@ public void shouldThrowOnTypeMismatchCannotCoerce() {
     assertThat(e.getMessage(), containsString("Expected type"));
   }
 
+  @Test
+  public void shouldThrowOnInsertPartition() {
```
Comment (vcrfxia) [inline ksqldb-engine/src/test/java/io/confluent/ksql/engine/generic/GenericRecordFactoryTest.java:368]:
```suggestion
  public void shouldThrowOnInsertRowpartition() {
```

File: ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/insert-values.json [hunk 10]
```diff
@@ -431,6 +431,66 @@
...
       ]
+    },
+    {
+      "name": "verify INSERT VALUES fails for ROWPARTITION (matching types)",
+      "properties": {
+        "ksql.rowpartition.rowoffset.enabled": true
+      },
+      "statements": [
+        "CREATE STREAM input (id INT KEY, val STRING) WITH (kafka_topic='test_topic', value_format='JSON');",
+        "INSERT INTO input (ROWPARTITION) VALUES (5);"
```
Comment (vcrfxia) [inline ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/insert-values.json:442]: nit: add more columns into the INSERT VALUES statement before ROWPARTITION.

File: ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/insert-values.json [hunk 11]
```diff
@@ -431,6 +431,66 @@
...
+        "INSERT INTO input (ROWOFFSET) VALUES (5);"
+      ],
+      "expectedError": {
+        "type": "io.confluent.ksql.rest.entity.KsqlStatementErrorMessage",
+        "message": "Inserting into column `ROWOFFSET` is not allowed.",
+        "status": 400
+      }
+    },
+    {
+      "name": "verify INSERT VALUES fails for ROWPARTITION (non-matching types)",
```
Comment (vcrfxia) [inline ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/insert-values.json:466]: I don't think this test (and the one below) are necessary.

File: ksqldb-engine/src/main/java/io/confluent/ksql/engine/generic/GenericRecordFactory.java (hunk 5 above)
Comment (Sullivan-Patrick) [reply ksqldb-engine/src/main/java/io/confluent/ksql/engine/generic/GenericRecordFactory.java:85]: I agree, I'll make the method accept any column name.

File: ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/insert-values.json (hunk 11 above)
Comment (Sullivan-Patrick) [reply ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/insert-values.json:466]: I was on the fence on those, SGTM.

File: ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/insert-values.json [hunk 12]
```diff
@@ -439,6 +439,8 @@
       },
       "statements": [
         "CREATE STREAM input (id INT KEY, val STRING) WITH (kafka_topic='test_topic', value_format='JSON');",
+        "INSERT INTO input (id VALUES (5);",
```
Comment (vcrfxia) [inline ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/insert-values.json:442]: I think these changes are a misunderstanding of one of my previous comments -- my suggestion was to add more columns to the single row being inserted, not to add more rows.

File: ksqldb-engine/src/main/java/io/confluent/ksql/engine/generic/GenericRecordFactory.java (hunk 5 above)
Comment (vcrfxia) [reply ksqldb-engine/src/main/java/io/confluent/ksql/engine/generic/GenericRecordFactory.java:85]: We don't need this anymore, in light of the update you made to `isDisallowedForInsertValues()`.

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/row-offset-and-partition.json [hunk 13]
```diff
@@ -19,13 +19,14 @@
         "sources": [
           {"name": "OUTPUT", "type": "stream", "schema": "ID INT KEY, ROWPARTITION STRING, ROWOFFSET STRING"}
         ]
-      }
+      },
+      "comments": "Should be removed after rowpartition.rowoffset.enabled feature flag is removed"
```
Comment (vcrfxia) [inline ksqldb-functional-tests/src/test/resources/query-validation-tests/row-offset-and-partition.json:23]: nit: it's nice to have "comments" above "name", else the comments are easily missed.

File: ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/insert-values.json [hunk 14]
```diff
@@ -431,6 +431,36 @@
       "outputs": [
         {"topic": "test_topic", "timestamp": 1234, "key": {"F1": "key"}, "value": {"ID": 10}}
       ]
+    },
+    {
+      "name": "verify INSERT VALUES fails for ROWPARTITION (matching types)",
```
Comment (vcrfxia) [inline ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/insert-values.json:436]:
```suggestion
      "name": "verify INSERT VALUES fails for ROWPARTITION",
```
and similarly for the test below -- it's unclear to readers what "matching types" means and I don't think it adds much.

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/row-offset-and-partition.json [hunk 15]
```diff
@@ -73,6 +76,89 @@
...
+      "statements": [
+        "CREATE STREAM input (id INT KEY, ROWPARTITION string, ROWOFFSET string) WITH (kafka_topic='test_topic', value_format='JSON');"
+      ],
+      "expectedException": {
+        "type": "io.confluent.ksql.util.KsqlStatementException",
+        "message": "'ROWPARTITION' is a reserved column name. You cannot use it as a name for a column."
+      }
+    },
+    {
+      "name": "test ROWOFFSET and ROWPARTITION source user columns with pseudocolumn data types- ROWOFFSET/ROWPARTITION enabled",
```
Comment (vcrfxia) [inline ksqldb-functional-tests/src/test/resources/query-validation-tests/row-offset-and-partition.json:124]: This is great test coverage but it looks unrelated to the main changes in this PR (for INSERT VALUES). Might make more sense as a separate PR? That said, if the build passes and you want to keep these new tests in this PR, that's fine too.

**Review by vcrfxia — COMMENTED** [review]
Thanks @Sullivan-Patrick -- LGTM besides the minor things inline. Let's merge once the build passes.

**Review by vcrfxia — APPROVED** [review]
Thanks @Sullivan-Patrick !
